dev/TabView/TabView.cpp

```diff
@@ -1232,18 +1297,131 @@ int TabView::GetItemCount()
     }
 }
 
-bool TabView::SelectNextTab(int increment)
+bool TabView::MoveFocus(bool moveForward)
 {
-    bool handled = false;
-    const int itemsSize = GetItemCount();
-    if (itemsSize > 1)
+    auto focusedControl = winrt::FocusManager::GetFocusedElement() ? winrt::FocusManager::GetFocusedElement().try_as<winrt::Control>() : nullptr;
+
+    // If there's no focused control, then we have nothing to do.
+    if (!focusedControl)
     {
-        auto index = SelectedIndex();
-        index = (index + increment + itemsSize) % itemsSize;
-        SelectedIndex(index);
-        handled = true;
+        return false;
     }
-    return handled;
+
+    // Focus goes in this order:
+    //
+    //    Tab 1 -> Tab 1 close button -> Tab 2 -> Tab 2 close button -> ... -> Tab N -> Tab N close button -> Add tab button -> Tab 1
+    //
+    // Any element that's not focusable is skipped.
+    //
+    std::vector<winrt::Control> focusOrderList;
+
+    for (int i = 0; i < GetItemCount(); i++)
+    {
+        if (auto tab = ContainerFromIndex(i).try_as<winrt::TabViewItem>())
+        {
+            if (SharedHelpers::IsFocusable(tab, false /* checkTabStop */))
+            {
+                focusOrderList.push_back(tab);
+
+                if (auto closeButton = winrt::get_self<TabViewItem>(tab)->GetCloseButton())
+                {
+                    if (SharedHelpers::IsFocusable(closeButton, false /* checkTabStop */))
+                    {
+                        focusOrderList.push_back(closeButton);
+                    }
+                }
+            }
+        }
+    }
+
+    if (auto&& addButton = m_addButton.get())
+    {
+        if (SharedHelpers::IsFocusable(addButton, false /* checkTabStop */))
+        {
+            focusOrderList.push_back(addButton);
+        }
+    }
+
+    auto position = std::find(focusOrderList.begin(), focusOrderList.end(), focusedControl);
+
+    // The focused control is not in the focus order list - nothing for us to do here either.
+    if (position == focusOrderList.end())
+    {
+        return false;
+    }
+
+    // At this point, we know that the focused control is indeed in the focus list, so we'll move focus to the next or previous control in the list.
+
+    int sourceIndex = static_cast<int>(position - focusOrderList.begin());
+    const int listSize = static_cast<int>(focusOrderList.size());
+    const int increment = moveForward ? 1 : -1;
+    int currentIndex = sourceIndex + increment;
+
+    while (currentIndex != sourceIndex)
```

Comment:
Can you help me understand why this needs to be a loop? We know what element we are moving focus to, so why do we need to do things in a loop?

Reply:
control.Focus() can return false.  It *shouldn't*, but the intention was to handle this in the case that it does.  Might be better to just assert, though.  I can change that.